src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -9,6 +9,336 @@ namespace System.Numerics.Tensors
+            {
+                ref float xRef = ref MemoryMarshal.GetReference(x);
+                ref float yRef = ref MemoryMarshal.GetReference(y);
+
+                Vector256<float> dotProductVector = Vector256<float>.Zero;
+                Vector256<float> xSumOfSquaresVector = Vector256<float>.Zero;
+                Vector256<float> ySumOfSquaresVector = Vector256<float>.Zero;
+
+                // Process vectors, summing their dot products and squares, as long as there's a vector's worth remaining.
+                int oneVectorFromEnd = x.Length - Vector<float>.Count;
+                do
+                {
+                    Vector256<float> xVec = Vector256.LoadUnsafe(ref xRef, (uint)i);
+                    Vector256<float> yVec = Vector256.LoadUnsafe(ref yRef, (uint)i);
+
+                    dotProductVector += (xVec * yVec);
+                    xSumOfSquaresVector += (xVec * xVec);
+                    ySumOfSquaresVector += (yVec * yVec);
+
+                    i += Vector256<float>.Count;
+                }
+                while (i <= oneVectorFromEnd);
+
+                // Sum the vector lanes into the scalar result.
+                dotProduct += Vector256.Sum(dotProductVector);
+                xSumOfSquares += Vector256.Sum(xSumOfSquaresVector);
+                ySumOfSquares += Vector256.Sum(ySumOfSquaresVector);
+            }
+            else if (Vector128.IsHardwareAccelerated && x.Length >= Vector128<float>.Count)
+            {
+                ref float xRef = ref MemoryMarshal.GetReference(x);
+                ref float yRef = ref MemoryMarshal.GetReference(y);
+
+                Vector128<float> dotProductVector = Vector128<float>.Zero;
+                Vector128<float> xSumOfSquaresVector = Vector128<float>.Zero;
+                Vector128<float> ySumOfSquaresVector = Vector128<float>.Zero;
+
+                // Process vectors, summing their dot products and squares, as long as there's a vector's worth remaining.
+                int oneVectorFromEnd = x.Length - Vector<float>.Count;
+                do
+                {
+                    Vector128<float> xVec = Vector128.LoadUnsafe(ref xRef, (uint)i);
+                    Vector128<float> yVec = Vector128.LoadUnsafe(ref yRef, (uint)i);
+
+                    dotProductVector += (xVec * yVec);
+                    xSumOfSquaresVector += (xVec * xVec);
+                    ySumOfSquaresVector += (yVec * yVec);
+
+                    i += Vector128<float>.Count;
+                }
+                while (i <= oneVectorFromEnd);
+
+                // Sum the vector lanes into the scalar result.
+                dotProduct += Vector128.Sum(dotProductVector);
+                xSumOfSquares += Vector128.Sum(xSumOfSquaresVector);
+                ySumOfSquares += Vector128.Sum(ySumOfSquaresVector);
+            }
+
+            // Process any remaining elements past the last vector.
+            for (; (uint)i < (uint)x.Length; i++)
```

Comment:
```suggestion
            for (; i < x.Length; i++)
```
Are the casts needed here? Index access below is with `int`, so I don't see any benefit from the `uint` in the loop.

Reply:
Technically the cast on the right shouldn't be necessary, but that's only a recent development, and so most of our code does so, and this is just staying consistent with that until they're all cleaned up. On the left hand side, unless the JIT can prove `i` isn't negative, it's necessary to enable it to avoid a bounds check (although there's another issue that prevents it from doing so with some loop shapes).